packages/flutter_test/lib/src/finders.dart

```diff
@@ -460,57 +483,443 @@ class CommonFinders {
+  /// Runs the given callback using cached results.
+  ///
+  /// While in this callback, this [FinderBase] will cache the results from the
+  /// next call to [find] or [tryFind] and then no longer evaluate new results
+  /// until the callback completes. After the first call, all calls to [find],
+  /// [tryFind] or [found] will return the same results without evaluating.
+  void runCached(VoidCallback run) {
+    reset();
+    _cached = true;
+    try {
+      run();
+    } finally {
+      reset();
+      _cached = false;
+    }
+  }
+
+  /// Resets all state of this [FinderBase].
+  ///
+  /// Generally used between tests to reset the state of [found] if a finder is
+  /// used across multiple tests.
+  void reset() {
+    _found = null;
+  }
+
+  /// A string representation of this finder or its results.
+  ///
+  /// By default, this describes the results of the search in order to play
+  /// nicely with [expect] and its output when a failure occurs. If you wish
+  /// to get a string representation of the finder itself, pass [describeSelf]
+  /// as `true`.
+  @override
+  String toString({bool describeSelf = false}) {
+    if (describeSelf) {
+      return 'A finder that searches for ${describeMatch(Plurality.many)}.';
+    } else {
+      if (!hasFound) {
+        find();
+      }
+      return found.toString();
+    }
+  }
+}
+
+/// The results of searching with a [FinderBase].
+class FinderResult<CandidateType> extends Iterable<CandidateType> {
+  /// Creates a new [FinderResult] that describes the [values] using the given
+  /// [describeMatch] callback.
+  ///
+  /// {@macro flutter_test.finders.FinderBase.describeMatch}
+  FinderResult(DescribeMatchCallback describeMatch, Iterable<CandidateType> values)
+    : _describeMatch = describeMatch, _values = values;
+
+  final DescribeMatchCallback _describeMatch;
+  final Iterable<CandidateType> _values;
+
+  @override
+  Iterator<CandidateType> get iterator => _values.iterator;
+
+  @override
+  String toString() {
+    final List<CandidateType> valuesList = _values.toList();
+    // This will put each value on its own line with a comma and indentation
+    final String valuesString = valuesList.fold(
+      '',
+      (String current, CandidateType candidate) => '$current\n  $candidate,',
     );
+    return 'Found ${valuesList.length} ${_describeMatch(Plurality._fromNum(valuesList.length))}: ['
+      '${valuesString.isNotEmpty ? '$valuesString\n' : ''}'
+      ']';
   }
+}
+
+/// Provides backwards compatibility with the original [Finder] API.
+// TODO(pdblasi-google): Deprecate methods on this mixin.
+mixin _LegacyFinderMixin on FinderBase<Element> {
+  Iterable<Element>? _precacheResults;
+
+  /// Whether this finder skips nodes that are offstage.
+  ///
+  /// If this is true, then the elements are walked using
+  /// [Element.debugVisitOnstageChildren]. This skips offstage children of
+  /// [Offstage] widgets, as well as children of inactive [Route]s.
+  bool get skipOffstage;
```

Comment:
Why is this part of the legacy? What's the replacement for skipOffstage?

Reply:
It _was_ when I was also going to replace `skipOffstage` to have it follow our "default to false" and naming schemes better, but when I dropped the new property I forgot to move this. Good catch!

Reply:
Moved.